src/foam/lib/csv/CSV.js

```diff
@@ -0,0 +1,256 @@
+
+  properties: [
+    {
+      class: 'String',
+      name: 'buf_',
+      value: ''
+    },
+    {
+      class: 'String',
+      name: 'delimiter',
+      value: ','
+    },
+    {
+      class: 'String',
+      name: 'nestedObjectSeperator',
+      value: '__'
+    },
+    {
+      class: 'String',
+      name: 'nlStr',
+      value: '\n'
+    },
+    {
+      class: 'String',
+      name: 'undefinedStr',
+      value: ''
+    },
+    {
+      class: 'Boolean',
+      name: 'outputHeaderRow',
+      value: true
+    },
+    {
+      class: 'Function',
+      name: 'propertyPredicate',
+      value: function(o, p) { return ! p.transient; }
+    }
+  ],
+
+  methods: [
+    function out() {
+      for ( var i = 0 ; i < arguments.length ; i++ ) { 
+        this.buf_ += arguments[i];
+      }
+
+      return this;
+    },
+
+    function toCSV(o) {
+      // Resets buffer
+      this.reset();
+
+      // Outputs object headers
+      this.outputHeader(o);
+
+      // Outputs object values
+      this.output(o, true);
+      this.out(this.nlStr);
+
+      return this.buf_;
+    },
+
+    function outputHeader(o) {
+      if ( ! this.outputHeaderRow ) return;
+
+      this.outputPropertyName(o, '', true);
+      this.out(this.nlStr);
+    },
+
+    function outputHeaderTitle(o, prefix, first) {
+      this.out(first ? '' : this.delimiter)
+          .out(this.escapeString(prefix + o));
+    },
+
+    function outputPropertyName_(o, p, prefix, first) {
+      if ( ! this.propertyPredicate(o, p) ) return;
+
+      // Checks if property is enum, or object with properties of its own
+      if ( p.of && ( ! foam.core.AbstractEnum.isInstance( o[p.name] ) ) ) {
+        // Gets new empty object if FObjectProperty is currently undefined
+        // Done to permit appropriate headers for multi-line CSVs (multiple objects to convert)
+        if ( o[p.name] == undefined ) o[p.name] = foam.lookup(p.of.id);
+
+        // Appends object name to prefix for CSV Header
+        prefix += p.name + this.nestedObjectSeperator;
+        this.outputPropertyName(o[p.name], prefix, first);
+      } else {
+        this.outputPropertyName(p.name, prefix, first);
+      }
+    },
+
+    {
+      name: 'outputPropertyName',
+      code: foam.mmethod({
+        FObject:   function(o, prefix, first) {
+          // Gets and recurses through object properties
+          var ps = o.cls_.getAxiomsByClass(foam.core.Property);
+
+          for ( var i = 0 ; i < ps.length ; i++ ) {
+            this.outputPropertyName_(o, ps[i], prefix, (i == 0 && first));
+          }
+        },
+        Array: function(o) { /* Ignore arrays in CSV */ },
+        Function: function(o) { /* Ignore functions in CSV */ },
+        Object: function(o) { /* Ignore generic objects in CSV */ }
+      }, function(o, prefix, first) { this.outputHeaderTitle(o, prefix, first); })
+    },
+
+    function outputProperty(o, p, first) {
+      if ( ! this.propertyPredicate(o, p) ) return;
```

Comment:
A little more straight-forward, and less code, to just do:
if ( this.propertyPredicate(o, p) ) this.output(o[p.name], first);